File: src/main/java/ru/notes/view/MainView.java
package ru.notes.view;

import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.grid.Grid;
import com.vaadin.flow.component.icon.VaadinIcon;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;
import com.vaadin.flow.component.textfield.TextField;
import com.vaadin.flow.data.value.ValueChangeMode;
import com.vaadin.flow.router.Route;
import org.springframework.beans.factory.annotation.Autowired;
import ru.notes.component.NoteEditor;
import ru.notes.model.Note;
import ru.notes.repository.NoteRepository;

@Route("")
public class MainView extends VerticalLayout {
    private final NoteRepository noteRepository;

    private final NoteEditor noteEditor;

    private final TextField filter = new TextField();
    private final Button addNewNoteButton = new Button("New note", VaadinIcon.PLUS.create());
    // build layout
    private final HorizontalLayout toolbar = new HorizontalLayout(filter, addNewNoteButton);
    private Grid<Note> grid = new Grid<>(Note.class);

    @Autowired
    public MainView(NoteRepository noteRepository, NoteEditor noteEditor) {
        this.noteRepository = noteRepository;
        this.noteEditor = noteEditor;

        filter.setPlaceholder("filter");
        // Hook logic to components

        // Replace listing with filtered content when user changes filter
        filter.setValueChangeMode(ValueChangeMode.EAGER);
        filter.addValueChangeListener(field -> notesList(field.getValue()));

        add(toolbar, grid, noteEditor);
        // Connect selected Note to editor or hide if none is selected
        grid
                .asSingleSelect()
                .addValueChangeListener(e -> noteEditor.editNote(e.getValue()));
        // Instantiate and edit new Note the new button is clicked
        addNewNoteButton.addClickListener(e -> noteEditor.editNote(new Note()));
        // Listen changes made by the editor, refresh data from backend
        noteEditor.setChangeHandler(() -> {
            noteEditor.setVisible(false);
            notesList(filter.getValue());
        });
        // Initialize listing
        notesList("");
    }

    private void notesList(String name) {
        if (name.isEmpty()) {
            grid.setItems(this.noteRepository.findAll());
        } else {
            grid.setItems(this.noteRepository.findByName(name));
        }
    }
//    private void notesList(String name) {
//        if (name.isEmpty()) {
//            grid.setItems(this.noteService.getAllNotes());
//        } else {
//            grid.setItems(this.noteService.findNotes(name));
//        }
//    }
}

File: src/main/kotlin/ru/notes/view/MainView.kt
package ru.notes.view

import com.vaadin.flow.component.button.Button
import com.vaadin.flow.component.grid.Grid
import com.vaadin.flow.component.icon.VaadinIcon
import com.vaadin.flow.component.orderedlayout.HorizontalLayout
import com.vaadin.flow.component.orderedlayout.VerticalLayout
import com.vaadin.flow.component.textfield.TextField
import com.vaadin.flow.data.value.ValueChangeMode
import com.vaadin.flow.router.Route
import ru.notes.component.NoteEditor
import ru.notes.model.Note
import ru.notes.repository.NoteRepository


@Route("")
class MainView(private val noteRepository: NoteRepository,
               private val noteEditor: NoteEditor
) : VerticalLayout() {
    private val filter = TextField()
    private val addNewNoteButton = Button("New note", VaadinIcon.PLUS.create())
    private val toolbar = HorizontalLayout(filter, addNewNoteButton)
    private val grid = Grid(Note::class.java)

    init {
        initFilter()
        editNoteListener() //connect selected line to editor or hide if none is selected
        addNoteListener()  //edit new Note the new button is clicked
        refreshDataFromBackend()
        listNotes("")
        add(toolbar, grid, noteEditor)
    }

    private fun refreshDataFromBackend() {
        noteEditor.setChangeHandler(object : NoteEditor.ChangeHandler {
            override fun onChange() {
                noteEditor.isVisible = false
                listNotes(filter.value)
            }
        })
    }

    private fun addNoteListener() {
        addNewNoteButton.addClickListener {
            noteEditor.editNote(Note(null, null, null))
        }
    }

    private fun editNoteListener() {
        grid.asSingleSelect()
                .addValueChangeListener {
                    noteEditor.editNote(it.value)
                }
    }

    private fun initFilter() {
        filter.placeholder = "filter"
        filter.valueChangeMode = ValueChangeMode.EAGER
        filter.addValueChangeListener {
            listNotes(it.value)
        }
    }

    private fun listNotes(name: String) {
        grid.setItems(
                if (name.isEmpty())
                    noteRepository.findAll() else
                    noteRepository.findByName(name)
        )
    }
}